### ISSUE TITLE
Patch to allow image sizes to be specified in "in", "mm", when using --browserwidth {high_dpi}

### ISSUE THREAD

**Issue by michaelrsweet** [michaelrsweet/htmldoc#289]
Version: 1.8-current
Original reporter: Iain W. Bird

When using very high browserwidth options, e.g. to allow high DPI background images:

htmldoc --browserwidth 3307 --webpage test2.htm --outfile test2.pdf --size A4 --jpeg=70 --no-title

images become very small unless specified in percent.

This patch allows image sizes to be specified in "in", "mm" or "px" (96DPI).  If units are not specified, the pre-existing behaviour is retained.

Use in HTML as follows, e.g.

<h1>TEST IMAGE SIZE NATIVE</h1>
<img src="./tulips.jpg">
<br />
<h1>TEST IMAGE SIZE 80%</h1>
<img src="./tulips.jpg" width="80%">
<br />
<h1>TEST IMAGE SIZE 1 INCH</h1>
<img src="./tulips.jpg" width="1in" height="1in">
<table width="1in" border="1">
<tr><td >TABLE WIDTH 1in</td></tr>
</table>
<br />
<h1>TEST IMAGE SIZE 2x1 INCH</h1>
<img src="./tulips.jpg" width="2in" height="1in">
<table width="2in" border="1">
<tr><td >TABLE WIDTH 2in</td></tr>
</table>
<br />
<h1>TEST IMAGE HEIGHT 50mm</h1>
<img src="./tulips.jpg" height="50mm">
<br />
<h1>TEST IMAGE WIDTH 144pt</h1>
<img src="./tulips.jpg" width="144pt" >
<br />
<h1>TEST IMAGE WIDTH 12pc</h1>
<img src="./tulips.jpg" width="12pc" >
<br />
<h1>TEST IMAGE WIDTH 530px</h1>
<img src="./tulips.jpg" width="530px">
<br />
<h1>TEST IMAGE WIDTH 530</h1>
<img src="./tulips.jpg" width="530">
<br />

Patch file has been tested under cygwin and MS-VC 2013 for the htmldoc executable

**Comment by michaelrsweet** [comment]
Might consider supporting the style attribute and width/height properties instead, otherwise the HTML input won't be valid...

**Comment by birdwes** [comment]
Hi Michael,

The reason I implemented in my local copy this was for compatibility with Internet Explorer print templates as described here: [](https://msdn.microsoft.com/en-us/library/bb250460) . I use an embedded IE as an engine to print the page if there is no email address.  If there is an email address, it is converted to PDF and sent.  Internet Explorer accepts "in","mm","px" etc, even though it is not really standards compliant.  

Thanks for your good work.  

Iain W. Bird

**Comment by birdwes** [comment]
https://msdn.microsoft.com/en-us/library/bb250460

**Comment by michaelrsweet** [comment]
Yeah, well I *do* want to keep things standards-compliant as much as possible, particularly as I implement support for EPUB which requires strict XHTML conformance...

Like I said, I'm considering adding support for things like:

    <img style="width: 4in; height: 3in;" src="..." ...>

to override the normal browser width/PPI calculations *and* remain standards-compliant.
